### PR TITLE
Mean mechanism implementation

### PR DESCRIPTION
**Pull request by dvadym**
This PR implements `MeanMechanism` class (in dp_computations.py). This class is used for computing DP Mean in `MeanCombiner`.  This is the part of the refactoring for making adding noise in more OOP style. This refactoring is needed for supporting PLD and l1 sensitivities in future.

Additional possibility which adds this refactoring it's possible to have more detailed description in explain computation reports about computin DP mean, here is an example:

 _5. DP mean computation: 
    a. Computed 'normalized_sum' = sum of (value - 3.0)
    b. Applied to 'count' Laplace mechanism:  parameter=6.0  eps=0.3333333333333333  l1_sensitivity=2.0
    c. Applied to 'normalized_sum' Laplace mechanism:  parameter=12.0  eps=0.3333333333333333  l1_sensitivity=4.0_

### CODE REVIEW

File: pipeline_dp/dp_computations.py [hunk 1]
```diff
@@ -529,6 +536,44 @@ def describe(self) -> str:
                 f"l2_sensitivity={self.sensitivity}")
 
 
+class MeanMechanism:
+    """Computes DP mean.
+
+    It computes DP mean as a ratio of DP sum and DP count. For improving
+    utility the normalization to [min_value, max_value] is performed.
+    It works in the following way:
+    1. normalized_sum = \sum(x_i-mid), where mid = (min_value+max_value)/2.
```
Comment (miracvbasaran) [inline pipeline_dp/dp_computations.py:545]: Does Python support Latex?

File: pipeline_dp/dp_computations.py [hunk 2]
```diff
@@ -529,6 +536,44 @@ def describe(self) -> str:
                 f"l2_sensitivity={self.sensitivity}")
 
 
+class MeanMechanism:
+    """Computes DP mean.
+
+    It computes DP mean as a ratio of DP sum and DP count. For improving
+    utility the normalization to [min_value, max_value] is performed.
```
Comment (miracvbasaran) [inline pipeline_dp/dp_computations.py:543]: Isn't it rather normalization to `mid_value`?

File: pipeline_dp/dp_computations.py [hunk 3]
```diff
@@ -529,6 +536,44 @@ def describe(self) -> str:
...
+
+    def __init__(self, range_middle: float, count_mechanism: AdditiveMechanism,
+                 sum_mechanism: AdditiveMechanism):
+        self._range_middle = range_middle
+        self._count_mechanism = count_mechanism
+        self._sum_mechanism = sum_mechanism
+
+    def compute_mean(self, count: int, normalized_sum: float):
+        dp_count = self._count_mechanism.add_noise(count)
+        denominator = max(1.0, dp_count)  # to avoid division on a small number.
```
Comment (miracvbasaran) [inline pipeline_dp/dp_computations.py:563]: I think it's just to avoid division by 0, no?

File: pipeline_dp/combiners.py [hunk 4]
```diff
@@ -396,10 +426,26 @@ def metrics_names(self) -> List[str]:
         return self._metrics_to_compute
 
     def explain_computation(self) -> ExplainComputationReport:
-        return lambda: f"Computed mean with (eps={self._params.eps} delta={self._params.delta})"
+        return lambda: f"DP mean computation: \n" + self.get_mechanism(
```
Comment (miracvbasaran) [inline pipeline_dp/combiners.py:429]: nit: whitespace before `\n` might be reduntant

File: tests/combiners_test.py [hunk 5]
```diff
@@ -111,9 +111,21 @@ def test_create_compound_combiner(self, metrics, expected_combiner_types):
...
-        mock_budgets = [
-            f"budget{i}" for i in range(len(expected_combiner_types))
-        ]
+        n_budget_requests = len(expected_combiner_types)
+        if dp_combiners.MeanCombiner in expected_combiner_types:
+            n_budget_requests += 1
+        mock_budgets, expected_budgets = [], []
+        for i, expeced_type in enumerate(expected_combiner_types):
+            if expeced_type == dp_combiners.MeanCombiner:
+                # MeanCombiner requests budgets twice, for Count and for Sum.
```
Comment (miracvbasaran) [inline tests/combiners_test.py:120]: super-nit: `one for Count and one for Sum`

File: tests/combiners_test.py [hunk 6]
```diff
@@ -451,28 +463,21 @@ def test_compute_metrics_no_noise(self):
...
+        count, normalized_sum = 100, 150
+
+        # Set mock noise to be a deterministic addition of 1, which can be
+        # easily tested.
+        mock_add_noise.side_effect = lambda x: x + 1
+
+        output = combiner.compute_metrics((count, normalized_sum))
+        self.assertEqual(output["count"], 101)  # add_noise(100)
+        # expected_sum = add_noise(normalized_sum) + mid_range*noisy_count = 353
+        # More details are in dp_computations.MeanMechanism docsting.
```
Comment (miracvbasaran) [inline tests/combiners_test.py:478]: nit: `docstring`

File: tests/dp_computations_test.py [hunk 7]
```diff
@@ -766,6 +766,70 @@ def test_compute_sensitivities_for_sum_min_max_per_partition(self):
...
+
+    @patch("pipeline_dp.dp_computations.GaussianMechanism.add_noise")
+    @patch("pipeline_dp.dp_computations.LaplaceMechanism.add_noise")
+    def test_compute_mean(self, mock_sum_add_noise, mock_count_add_noise):
+        mechanism = self.create_mean_mechanism()
+
+        mock_count_add_noise.return_value = 10  # dp_count
+        mock_sum_add_noise.return_value = 20  # dp_sum
+
+        expected_mean = 7  # range_middle + dp_sum/dp_count = 5 + 20/10
```
Comment (miracvbasaran) [inline tests/dp_computations_test.py:811]: `mid_range` instead of `range_middle` for consistency. Or the other way around.

(Also in other places)

**Review by miracvbasaran — APPROVED** [review]
Looks great, thanks!

File: pipeline_dp/dp_computations.py (hunk 3 above)
Comment (dvadym) [reply pipeline_dp/dp_computations.py:563]: I think that the problem is broader than division on 0, e.g. with division by 1e-10 we will have a huge number

File: pipeline_dp/dp_computations.py (hunk 1 above)
Comment (dvadym) [reply pipeline_dp/dp_computations.py]: Python comments are not treated as Latex :(. I've removed "\\".

File: pipeline_dp/dp_computations.py (hunk 2 above)
Comment (dvadym) [reply pipeline_dp/dp_computations.py]: Thanks, I've updated the docstring

File: tests/combiners_test.py (hunk 5 above)
Comment (dvadym) [reply tests/combiners_test.py]: thx

**Review by dvadym — COMMENTED** [review]
Thanks for review!

**Review by dvadym — COMMENTED** [review]
Thanks for review!

**Pull request merged**